modules/apps/remote-app/remote-app-admin-web/src/main/resources/com/liferay/remote/app/admin/web/internal/dependencies/js-portlet.js.tpl

```diff
@@ -0,0 +1,13 @@
+(function() {
```

Comment:
Is this really necessary? Couldn't we directly explore `<script type="module">`?

Reply:
In here I get the sense that we're mixing 2 things:
- A JS widget created with our tooling can be hosted elsewhere and act as a widget
- A JS function hosted elsewhere can act as a widget

While the first one is definitely desirable, we don't run the sources directly but after the build process, so it feels counterintuitive that you'd host them before building them if that was the case. I'd expect `js-toolkit` to produce already compiled amd modules to host externally if that's the case.

Reply:
Just to give some context

Yesterday, Iván and I talked about "parity" in terms of tooling: whatever we build with `js-toolkit` should be remotely deployable and accesible via URL pointing to the JS entry point. This does not prevent us from accepting other forms of remote JS but at least, if we're talking about JS Widget (TM) then it'd be desirable to use the same tooling regardless of how it gets deployed. Perhaps not all build tasks will be involved in each case but reuse is clearly something we want here.

We had some discussions about how much we can separate the registering and publication processes of remote URLs, and found out there were many possible combinations and that they are related (i.e. the shape of what's remotely deployed impacts on how it can be made available), which require decisions on our side about which things we'll support.

For this first iteration, our goal is that Brian accepts the pull with a really bare MVP, then evolve it. So, we're under the assumption that we'll have a JS function as the one we use for JS widgets (either created manually or with the tooling). If this gets merged, we'll have to see how much of the js widget build flow can be utilized in producing remote js widgets. That name should be easier to understand for end users as a particular kind of remote deployment, but to honor it, it has to be covered by the tooling.

Reply:
I explored it, but I think we are not ready yet because we probably want to migrate AMD loader to browser modules some time in the future (with the bundler 3/wepack 5 model: not exporting things by default, as opposed to bundler 2, it fits quite well). So, for now, I'm registering them under the AMD loader so that, in the future, we may swap everything at once.

The "problem" with using modules is that we would introduce another way of doing things (one more :sweat: ) for little gain (now). I prefer to defer it for the moment where we really take advantage of it instead of tying us to another scheme we may not unify in the future :grimacing:.

We investigated browser modules some time ago and it/us weren't mature enough to use it, but this time I feel we are very near to be capable of using them :crossed_fingers: . However, since the devil is in the details, and we always have bad luck with these feelings :sweat_smile: , I would not assume it yet.

Reply:
Our plan is to align remote js widgets with deployed js widgets (those coming from JS toolkit) so that we don't need to:

1. implement things twice
2. document things twice
3. * twice
4. probably very important: be able to route everything through the same place in the future. In the end, we would probably like to have JS widgets and treat them the same once they are registered with the portal, no matter how they came there...

Reply:
And the same for AMD/browser modules, custom elements, ...

Reply:
Once again, GH hid all comments but first before I posted my previous ones :tada: .

Reply:
Just to clarify, and please correct me if I'm wrong.

> whatever we build with js-toolkit should be remotely deployable and accesible via URL pointing to the JS entry point

What I think I see here is that you'd take the sources of a js-widget and upload them to a cloud storage. So, it's not "whatever you build", because you're not building it. Thus, you need to do source-modifications in-place, because on one hand you're dealing with sources whereas with the js toolkit we effectively build those into some type of bundle.

If the goal is to support "whatever we build", then I'd expect that to be already an AMD module ready to play within Liferay. So, when doing devops, you'd do `js-toolkit build && s3 upload build` rather than `s3 upload src` (pseudo code)

Reply:
My personal opinion is that we should support both things built with js-toolkit and things not built with anything (like the vanilla counter in the example).

If it were for me I wouldn't use js-toolkit at all, but as soon as people begin to ask for deduplication or "using portal's react and/or clay" we are gonna need some building, whether it is before the `s3 upload build` or during registration of the URL by transforming sources in the portal on-the-fly. 

Note that we have not yet done any transformation of sources in the portal on-the-fly, but theoretically it should be possible to do the same we do with the js-toolkit at that time (think of how Heroku fetches things from GH and builds them before deployment).

@jbalsas you seem to prefer the on-the-fly build in mind, is that right? or would we support both? :thinking:

Reply:
> @jbalsas you seem to prefer the on-the-fly build in mind, is that right? or would we support both? 🤔

I'm not sure I do, honestly.

```javascript
eval(source.replace('export default', 'module.exports='));
```

This was rather unexpected for me and the reason (other than the `fetch+eval` approach) why I brought it up. I would prefer no "on-the-fly" transformation of any kind whenever possible. 

I have the sense we'd be both breaking the principle of least astonishment (apparently my vanilla js file got turned into an amd module? 🤯 ) and get developers into an uncanney valley (I wrote an ES module and imported another one... why isn't this working as I expected?). 

I'd be happy if we could find a way to do less and be more explicit. If we're not doing a build step explicitly, can't we branch off the logic and try to stay closer to how you'd use the file on its own? If we can't, I'd rather keep it contained to a explicit output from the js-toolkit so you know what format to expect and can avoid doing weird source transformations.

Reply:
> apparently my vanilla js file got turned into an amd module

This is an internal implementation, not a contract. You could argue the same if we injected it as a `<script>` tag or if we had turned it into a browser module...

The fact is that you are deploying a JS file with a function inside and it is magically converted to a portlet. There has to be something in the middle. In other words: there doesn't exist any intuitive expectation for that (IMO).

Said that, I agree that I also thought about the `I wrote an ES module and imported another one` part and my idea was to provide a path for future development and be able to deploy everything as pure ES modules. But I agree that until we get there this may be misleading.

However, if we remove `export default` we need some way to tell the infrastructure what to look for inside the code, since there can be anything inside it (variables, functions, classes, etc...). How would we do that? :thinking:

Reply:
2> keep it contained to a explicit output from the js-toolkit

We also though about this, the problem is that in this case we cannot point to a single .js file, but we need to point to an "HTTP folder" (akin to the JAR file). It's not impossible, but diverts a lot from what Brian wants now, I think...

Reply:
> What I think I see here is that you'd take the sources of a js-widget and upload them to a cloud storage. So, it's not "whatever you build", because you're not building it.

I meant just that. I used "build" in a broader sense (not just packaging), so one would use the toolkit to create the project at the very least, then upload it somewhere. Still we're far from that as the entry point is just a function, but I guess this could be improved to accept more inputs such as the configuration.json.

Reply:
Let me investigate the browser module path a bit more... 

My fear when taking that route was to end up having two parallel namespaces for modules: one coming from the AMD loader, another from browser modules, that wasn't interoperable in the future. But maybe it's not the case, and we can begin by deploying these portlets as pure browser modules, then evolving the rest.

At least, let's give it a try.